### ISSUE TITLE
findbugs:NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE does not consider null validation

### ISSUE THREAD

**Issue by alixwar** [spotbugs/spotbugs#651]
**Problem**: A violation is triggered at the last line, saying that args may be null. This is clearly validated the line before (using org.apache.commons.lang3.Validate)

**Affected version**: 3.1.2 (SpotBugs)
**Seen in**: SonarQube (FindBugs plugin 3.7)

Reproducible code example (from an Android app):

```java
@Override
public void onCreate(@Nullable Bundle savedInstanceState) {
   super.onCreate(savedInstanceState);
   final Bundle args = getArguments();
   Validate.notNull(args);
   email = args.getString(ARG_EMAIL);
}
```

**Comment by npfistner** [comment]
**Problem:** Also if validated using java assert spotbugs still complains NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE

**Affected version:** 3.1.5 (SpotBugs)

```java
        assert entity.getId() != null;
        final long id = entity.getId().longValue();
```

**Comment by ThrawnCA** [comment]
I'm not sure that the `assert` keyword should be relied on, since assertions can be disabled. But the Commons Lang one is likely worth including; thanks for pointing it out.

**Comment by alixwar** [comment]
Same issue with `java.util.Objects.requireNonNull()`

**Comment by nmatt** [comment]
+1 for supporting `requireNonNull()`.

**Comment by gmbett** [comment]
Same issue with custom method that is annotated with @Contract("null -> false")

**Comment by alixwar** [comment]
I'm trying to add a reproducer in the SpotBugs project but I keep getting: 

```
The following classes needed for analysis were missing:
  org.apache.commons.lang.Validate
```

even after adding the dependency to the spotbugs-tests module 😕

**Comment by gmbett** [comment]
Any updates on this issue?

**Comment by victornoel** [comment]
@ThrawnCA any news on that?

**Comment by C-Otto** [comment]
There's a database of hardcoded method parameter annotations in https://github.com/spotbugs/spotbugs/blob/master/spotbugs/src/main/java/edu/umd/cs/findbugs/ba/DefaultNullnessAnnotations.java#L406

At the given line, the first argument of `Preconditions.checkNotNull` is assumed to be "non-null". As such, the analysis throws an exception of `null` is passed to `checkNotNull`.

The same holds true for other implementations, for example `Objects.requireNonNull`.

**Comment by marcopelegrini** [comment]
Oh man, this is so annoying... NPE is valid in some cases, therefore the use of `Objects .requireNonNull` ... or even if you do something like:

```
private String requireNonNull(@Nullable final String obj, final String message) {
        if (obj == null) {
            throw new IllegalArgumentException(message);
        }
        return obj;
    }
```

it still complains

**Comment by ByeBug** [comment]
any news on that and #456 ?

**Comment by hazendaz** [comment]
Fixed by #2709
